src/escalus_connection.erl

```diff
@@ -174,45 +183,119 @@ maybe_set_jid(Client = #client{props = Props}) ->
     end.
 
 -spec send(escalus:client(), exml_stream:element()) -> ok.
-send(#client{module = Mod, event_client = EventClient, rcv_pid = Pid, jid = Jid}, Elem) ->
+send(#client{module = Mod, event_client = EventClient, rcv_pid = Pid, jid = Jid} = Client, Elem) ->
     escalus_event:outgoing_stanza(EventClient, Elem),
     escalus_ct:log_stanza(Jid, out, Elem),
-    Mod:send(Pid, Elem).
+    Mod:send(Pid, Elem),
+    handle_stanza(Client, Elem, #{}, sent_stanza_handlers(Client)),
+    ok.
 
 -spec get_stanza(client(), any()) -> exml_stream:element().
-get_stanza(Conn, Name) ->
-    get_stanza(Conn, Name, ?TIMEOUT).
+get_stanza(Client, Name) ->
+    get_stanza(Client, Name, ?TIMEOUT).
 
 -spec get_stanza(client(), any(), timeout()) -> exml_stream:element().
 get_stanza(Client, Name, Timeout) ->
-    case get_stanza_safe(Client, Timeout) of
+    get_stanza(Client, Name, Timeout, none).
+
+-spec get_stanza(client(), any(), timeout(), filter_pred()) -> exml_stream:element().
+get_stanza(Client, Name, Timeout, Pred) ->
+    case get_stanza_safe(Client, Timeout, Pred) of
         {error, timeout} ->
             throw({timeout, Name});
         {Stanza, _} ->
             Stanza
     end.
 
--spec get_stanza_with_metadata(client(), any(), timeout()) ->
-    {exml_stream:element(), metadata()}.
+-spec get_stanza_with_metadata(client(), any(), timeout()) -> {exml_stream:element(), metadata()}.
 get_stanza_with_metadata(Client, Name, Timeout) ->
-    case get_stanza_safe(Client, Timeout) of
+    get_stanza_with_metadata(Client, Name, Timeout, fun(_) -> true end).
+
+-spec get_stanza_with_metadata(client(), any(), timeout(), filter_pred()) ->
+                                      {exml_stream:element(), metadata()}.
+get_stanza_with_metadata(Client, Name, Timeout, Pred) ->
+    case get_stanza_safe(Client, Timeout, Pred) of
         {error, timeout} ->
             throw({timeout, Name});
         StanzaWithMetadata ->
             StanzaWithMetadata
     end.
 
--spec get_stanza_safe(client(), timeout()) ->
-    {error, timeout} | {exml_stream:element(), map()}.
-get_stanza_safe(#client{rcv_pid = Pid, jid = Jid}, Timeout) ->
+-spec get_stanza_safe(client(), timeout()) -> {error, timeout} | {exml_stream:element(), metadata()}.
+get_stanza_safe(Client, Timeout) ->
+    get_stanza_safe(Client, Timeout, none).
```

Comment:
`none` could probably be removed by specifying a `Pred` which always returns `true` (if I understand it correctly).

Reply:
The original version of this code used `fun(_) -> true end` indeed. This was changed to support `none` to simplify the data types and use `filter_pred` everywhere. I can drop the support for `none` and create a new type e.g. `stanza_pred`.